enhancements/microshift/gateway-api-support.md

```diff
@@ -0,0 +1,434 @@
+not able to serve it yet. Since there are CRs in the same RPM that need CRDs
+to be available this could become a race condition when installing. The
+mitigation would be to turn the RPM into a controller with proper mechanisms
+to wait and sync resources. An RPM was chosen because of the dev preview nature
+of the feature, as it simplifies things on MicroShift.
+
+On the non-technical side, this feature is born as Dev Preview to wait on
+customer feedback and adoption. This has no mitigation other than help out
+potential users of the feature
+
+### Drawbacks
+The inclusion of OpenShift Service Mesh brings extra resource usage. As this is
+released as dev preview, users need to opt into it. Additional memory should be
+planned when enabling this feature, as OSSM components and the gateways
+themselves consume additional resources.
+
+## Open Questions [optional]
+N/A
+
+## Test Plan
+Gateway API support for MicroShift will be tested using existing test
+harness. A new suite will be created with a simple test for the overall setup
+with an HTTP route. Another test will ensure compatibility with the current
+OpenShift router deployment.
+
+## Graduation Criteria
+Gateway API for MicroShift is targeted to be Dev Preview on MicroShift 4.18.
+
+At the time of this writing OSSM is in Tech Preview state, expected to graduate
+to GA in 2025.
+
+Graduation criteria will be determined by customer adoption/feedback.
+
+### Dev Preview -> Tech Preview
+N/A
+
+### Tech Preview -> GA
+N/A
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+Gateway API and MicroShift RPMs get built from the same spec so they share
+version. Whenever an upgrade happens both of them are also upgraded at the
+same time.
+
+The bulk of manifests are CRDs which have their own versioning, so the
+upgrade/downgrade strategy relies on the components using them. Extracting the
+OSSM manifests from the bundle guarantees that Istio will be compatible with
+them.
+
+For Gateway and HTTPRoute/GRPCRoute resources, manual action may be needed but
+this is the application's responsibility.
+
+## Version Skew Strategy
+Building Gateway API and MicroShift RPMs from the same spec file means both
+should be updated together which means there should not be any version skew
+between them.
+
+## Operational Aspects of API Extensions
+N/A
+
+## Support Procedures
+In order to debug error situations with Gateway API and OSSM, Gateways and
+routes need to be inspected.
+
+The first thing to check is the gatewayclass status, it needs to be accepted
+and handled by the istio controller, as shown in the status:
+```
+$ oc get gatewayclass openshift-gateway-api -o yaml
+apiVersion: gateway.networking.k8s.io/v1
+kind: GatewayClass
+metadata:
+  creationTimestamp: "2024-10-14T07:44:20Z"
+  generation: 1
+  name: openshift-gateway-api
+  resourceVersion: "1060"
+  uid: 0707b37e-c3c6-4b8a-a344-24545d8b8351
+spec:
+  controllerName: openshift.io/gateway-controller
```

Comment:
This is the [controllerName when you're using Cluster Ingress](https://github.com/openshift/cluster-ingress-operator/blob/master/pkg/operator/controller/gatewayclass/controller.go#L31).  Does this work if Cluster Ingress is not installed on MicroShift?  It probably shouldn't.

Reply:
Yes, this is only a name for the GatewayClass that needs to be configured in Istio Pilot env var [PILOT_GATEWAY_API_CONTROLLER_NAME](https://istio.io/latest/docs/reference/commands/pilot-discovery/) and [Gateway](https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.GatewaySpec) CRs must use too.
Cluster ingress operator uses this to configure OSSM2 as well. Since cluster ingress operator is not present in MicroShift and in order to be as close to OCP as possible we chose the same name.